File: Python/Problems/Problems/problem2115.py
import sys
from typing import List, Dict

from .problem import Problem


class Problem2115(Problem):
    def test(self) -> bool:
        recipes = ["bread", "sandwich", "burger"]
        ingredients = [["yeast", "flour"], ["bread", "meat"], ["sandwich", "meat", "bread"]]
        supplies = ["yeast", "flour", "meat"]

        expected = ["bread", "burger", "sandwich"]

        result = self.findAllRecipes(recipes, ingredients, supplies)
        result.sort()

        print(','.join(result))

        return result == expected

    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        suppliesSet = set(supplies)
        addedRecipesSet = set()
        added = 1
        while added > 0:
            added = 0
            for i in range(len(recipes)):
                if recipes[i] in addedRecipesSet:
                    continue
                crafted = True
                for ingredient in ingredients[i]:
                    if ingredient not in suppliesSet:
                        crafted = False
                        break
                if crafted:
                    suppliesSet.add(recipes[i])
                    addedRecipesSet.add(recipes[i])
                    added += 1

        return list(addedRecipesSet)

File: C++/Problems/Problems/problem_2115.cpp
#include "problem_2115.h"

#include <unordered_set>

bool problem_2115::test() {
    std::vector<std::string> recipes { "bread", "sandwich", "burger" };
    std::vector<std::vector<std::string>> ingredients { {"yeast", "flour"},  {"bread", "meat"},  {"sandwich", "meat", "bread"}};
    std::vector<std::string> supplies  { "yeast", "flour", "meat" };

    std::vector<std::string> expected  { "burger", "sandwich", "bread" };

    auto result = findAllRecipes(recipes, ingredients, supplies);

    print1DVector(result);

    return result == expected;
}

std::vector<std::string> problem_2115::findAllRecipes(std::vector<std::string> &recipes,
    std::vector<std::vector<std::string>> &ingredients, std::vector<std::string> &supplies) {
    std::unordered_set suppliesSet(supplies.begin(), supplies.end());
    std::unordered_set<std::string> addedRecipesSet;
    int added = 1;
    while (added > 0)
    {
        added = 0;
        for (int i = 0; i < recipes.size(); ++i)
        {
            if (addedRecipesSet.contains(recipes[i]))
                continue;
            auto crafted = true;
            for (const auto& ingredient : ingredients[i])
            {
                if (!suppliesSet.contains(ingredient))
                {
                    crafted = false;
                    break;
                }
            }

            if (crafted)
            {
                suppliesSet.insert(recipes[i]);
                addedRecipesSet.insert(recipes[i]);
                added++;
            }
        }
    }

    return std::vector(addedRecipesSet.begin(), addedRecipesSet.end());
}
